Fix condition for weights_only unpickler for DTensor

mikaylagawarecki opened the pull request:
Same as #140739 but for DTensor (move safe globals for DTensor to `torch.distributed.tensor.__init__` and update error message to let user know `torch.distributed.tensor` must be imported to load DTensor)

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #140850
* __->__ #140740
* #140739

Differential Revision: [D65961690](https://our.internmc.facebook.com/intern/diff/D65961690)

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

mikaylagawarecki committed:
Fix condition for weights_only unpickler for DTensor

[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D65961690).

mikaylagawarecki committed:
Update on "Fix condition for weights_only unpickler for DTensor"

The additions should not have been gated on `if hasattr(torch.distributed, "dtensor")` 

With pytorch built with USE_DISTRIBUTED=1
```
>>> import torch
>>> hasattr(torch.distributed, "dtensor")
False
```


Differential Revision: [D65961690](https://our.internmc.facebook.com/intern/diff/D65961690)

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Fix condition for weights_only unpickler for DTensor"

The additions should not have been gated on `if hasattr(torch.distributed, "dtensor")` 

With pytorch built with USE_DISTRIBUTED=1
```
>>> import torch
>>> hasattr(torch.distributed, "dtensor")
False
```


Differential Revision: [D65961690](https://our.internmc.facebook.com/intern/diff/D65961690)

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D65961690).

mikaylagawarecki committed:
Update on "Fix condition for weights_only unpickler for DTensor"

Same as #140739 but for DTensor (move safe globals for DTensor to `torch.distributed.tensor.__init__` and update error message to let user know `torch.distributed.tensor` must be imported to load DTensor)


Differential Revision: [D65961690](https://our.internmc.facebook.com/intern/diff/D65961690)

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Fix condition for weights_only unpickler for DTensor"

Same as #140739 but for DTensor (move safe globals for DTensor to `torch.distributed.tensor.__init__` and update error message to let user know `torch.distributed.tensor` must be imported to load DTensor)


Differential Revision: [D65961690](https://our.internmc.facebook.com/intern/diff/D65961690)

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Fix condition for weights_only unpickler for DTensor"

Same as #140739 but for DTensor (move safe globals for DTensor to `torch.distributed.tensor.__init__` and update error message to let user know `torch.distributed.tensor` must be imported to load DTensor)


Differential Revision: [D65961690](https://our.internmc.facebook.com/intern/diff/D65961690)

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

malfet commented on test/distributed/_tensor/test_dtensor.py:
```diff
@@ -542,6 +546,51 @@ def test_dtensor_save_load(self):
+                import_string = (
+                    "import torch.distributed.tensor" if should_import else ""
+                )
+                script = f"""
+import torch
+{import_string}
+x = torch.load("{f.name}", weights_only=True)
+"""
```
Nit (no need for 3 line script, use semicolon as line separator. And no need to specify weigh_only as it is a default right now
```suggestion
                script = f"import torch;{import_sring};torch.load('{f.name}')"
```

malfet commented on test/distributed/_tensor/test_dtensor.py:
```diff
@@ -542,6 +546,51 @@ def test_dtensor_save_load(self):
+                        cwd=os.path.dirname(os.path.realpath(__file__)),
+                        stderr=subprocess.STDOUT,
+                    )
+                    if not should_import:
+                        raise RuntimeError(
+                            "Script executed successfully despite lack of import"
+                        )
```
How about something like that:
```python
cm = self.assertRaisesRegex(RuntimeError, "_picke.Unpickling...") else contextlib.nullcontext()
with cm:
    try:
      subprocess.check_call([sys.executable, "-c", script])
  except subprocess.CalledProcessError as e:
      raise RuntimeError(e.output.decode("utf-8"))
```
This way you can account for case when something should finish without an error and when something should raise an exception

malfet commented on test/distributed/_tensor/test_dtensor.py:
```diff
@@ -542,6 +546,51 @@ def test_dtensor_save_load(self):
+                try:
+                    subprocess.check_output(
+                        [sys.executable, "-c", script],
+                        cwd=os.path.dirname(os.path.realpath(__file__)),
```
Instead of specifying a working directory, why not path file via absolute path to the script?

mikaylagawarecki replied:
For this one actually added this following what was done here, copied over the comment as well

https://github.com/pytorch/pytorch/blob/175ba9fed6a718dd1d0add9c733ed49bab97cc6c/test/test_autograd.py#L10170-L10172

mikaylagawarecki committed:
Update on "Fix condition for weights_only unpickler for DTensor"

Same as #140739 but for DTensor (move safe globals for DTensor to `torch.distributed.tensor.__init__` and update error message to let user know `torch.distributed.tensor` must be imported to load DTensor)


Differential Revision: [D65961690](https://our.internmc.facebook.com/intern/diff/D65961690)

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Fix condition for weights_only unpickler for DTensor"

Same as #140739 but for DTensor (move safe globals for DTensor to `torch.distributed.tensor.__init__` and update error message to let user know `torch.distributed.tensor` must be imported to load DTensor)


Differential Revision: [D65961690](https://our.internmc.facebook.com/intern/diff/D65961690)

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Fix condition for weights_only unpickler for DTensor"

Same as #140739 but for DTensor (move safe globals for DTensor to `torch.distributed.tensor.__init__` and update error message to let user know `torch.distributed.tensor` must be imported to load DTensor)


Differential Revision: [D65961690](https://our.internmc.facebook.com/intern/diff/D65961690)

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Fix condition for weights_only unpickler for DTensor"

Same as #140739 but for DTensor (move safe globals for DTensor to `torch.distributed.tensor.__init__` and update error message to let user know `torch.distributed.tensor` must be imported to load DTensor)


Differential Revision: [D65961690](https://our.internmc.facebook.com/intern/diff/D65961690)

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

wz337 commented:
Thanks for helping to fix it!
